### PR TITLE
add i2c driver for gd32f450 MCU

### PR DESCRIPTION
**Pull request by GD32-MCU**
## Summary
Add I2C driver for GD32F450 MCU
## Testing
Has been  test on the GD32F450zk eval board

### CODE REVIEW

File: boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#define START_BLOCK 0
+
+#if BUFFSIZE>=CONFIG_AT24XX_MTD_BLOCKSIZE
+#  define NBLOCK    (BUFFSIZE/CONFIG_AT24XX_MTD_BLOCKSIZE)
+#else
+#  error "BUFFSIZE should bigger than CONFIG_AT24XX_MTD_BLOCKSIZE"
+#endif
+
+const uint8_t write_buf[BUFFSIZE] =
+                                    {
```
Comment (xiaoxiang781216) [inline boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c:67]: remove the extra space

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: I have removed the extra space , but there is still problem. Could you help me to check what else should be modified, Thanks.

Comment (xiaoxiang781216) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: run:
./tools/refresh.sh --silent gd32f450zk-eval:i2c0testnsh
and commit the update defconfig

Comment (xiaoxiang781216) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: @GD32-MCU Please remove the intermediate patch from PR.

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: Could you prompt me how to delete the intermediate patch？ When use git rebase, there is a status of "Your branch and 'origin/GD32F4' have diverged"
Should i delete all submissions and recommit again.

Comment (xiaoxiang781216) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: do you follow my previous step? I don't meet your situation before.

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: I am not follow the previous step, because of the rebase is fail.

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: Could i delete my branch and recreate it again, then i commit the changes again.

Comment (xiaoxiang781216) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: I fix your problem. If you need modify the patch again, please:
git checkout master
git branch -D GD32F4
git branch --track GD32F4 origin/GD32F4
git checkout GD32F4
do the modification
git add *
git commit --amend
git push -f origin

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: I will try it, thank you very much.

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: There is an error about "Build/macOS(sim-01)", could you help me to see why there is this error？
Thanks.

Comment (xiaoxiang781216) [reply boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_at24.c]: It's ci issue let's ignore it.

**Comment by xiaoxiang781216** [comment]
please squash the change into one commit

File: arch/arm/src/gd32f4/gd32f4xx_dma.c [hunk 2]
```diff
@@ -714,21 +714,17 @@ void gd32_dma_singlemode_setup(struct gd32_dma_channel_s *dmachan,
 
   regval |= init_struct->direction;
 
-  if (DMA_WIDTH_8BITS_SELECT == init_struct->periph_memory_width)
+  if (DMA_WIDTH_32BITS_SELECT == init_struct->periph_memory_width)
     {
-      regval |= DMA_MEMORY_WIDTH_8BIT;
+      regval |= (DMA_MEMORY_WIDTH_32BIT | DMA_PERIPH_WIDTH_32BIT);
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_dma.c:719]:
```suggestion
      regval |= DMA_MEMORY_WIDTH_32BIT | DMA_PERIPH_WIDTH_32BIT;
```

File: arch/arm/src/gd32f4/gd32f4xx_dma.c [hunk 3]
```diff
@@ -714,21 +714,17 @@ void gd32_dma_singlemode_setup(struct gd32_dma_channel_s *dmachan,
...
     }
   else if (DMA_WIDTH_16BITS_SELECT == init_struct->periph_memory_width)
     {
-      regval |= DMA_MEMORY_WIDTH_16BIT;
-    }
-  else if (DMA_WIDTH_32BITS_SELECT == init_struct->periph_memory_width)
-    {
-      regval |= DMA_MEMORY_WIDTH_32BIT;
+      regval |= (DMA_MEMORY_WIDTH_16BIT | DMA_PERIPH_WIDTH_16BIT);
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_dma.c:723]:
```suggestion
      regval |= DMA_MEMORY_WIDTH_16BIT | DMA_PERIPH_WIDTH_16BIT;
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 4]
```diff
@@ -0,0 +1,2731 @@
...
+/****************************************************************************
+ * Name: gd32_i2c_getreg
+ *
+ * Description:
+ *   Get a 16-bit register value by offset
+ *
+ ****************************************************************************/
+
+static inline uint16_t gd32_i2c_getreg(struct gd32_i2c_priv_s *priv,
+                                      uint8_t offset)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:603]:
```suggestion
static inline uint16_t gd32_i2c_getreg(struct gd32_i2c_priv_s *priv,
                                       uint8_t offset)
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 5]
```diff
@@ -0,0 +1,2731 @@
...
+/****************************************************************************
+ * Name: gd32_i2c_putreg
+ *
+ * Description:
+ *  Put a 16-bit register value by offset
+ *
+ ****************************************************************************/
+
+static inline void gd32_i2c_putreg(struct gd32_i2c_priv_s *priv,
+                                  uint8_t offset, uint16_t value)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:617]:
```suggestion
static inline void gd32_i2c_putreg(struct gd32_i2c_priv_s *priv,
                                   uint8_t offset, uint16_t value)
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 6]
```diff
@@ -0,0 +1,2731 @@
...
+ * Name: gd32_i2c_sem_init
+ *
+ * Description:
+ *   Initialize semaphores
+ *
+ ****************************************************************************/
+
+static inline void gd32_i2c_sem_init(struct gd32_i2c_priv_s *priv)
+{
+  nxsem_init(&priv->sem_excl, 0, 1);
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:876]: let's use `nxmutex`

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 7]
```diff
@@ -0,0 +1,2731 @@
...
+{
+  /* Reset the trace info for a new data collection */
+
+  priv->tndx       = 0;
+  priv->start_time = clock_systime_ticks();
+  gd32_i2c_traceclear(priv);
+}
+
+static void gd32_i2c_tracenew(struct gd32_i2c_priv_s *priv,
+                               uint32_t status)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:935]:
```suggestion
static void gd32_i2c_tracenew(struct gd32_i2c_priv_s *priv,
                              uint32_t status)
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 8]
```diff
@@ -0,0 +1,2731 @@
...
+  else
+    {
+      /* Just increment the count of times that we have seen this status */
+
+      trace->count++;
+    }
+}
+
+static void gd32_i2c_traceevent(struct gd32_i2c_priv_s *priv,
+                                 enum gd32_trace_e event, uint32_t parm)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:977]:
```suggestion
static void gd32_i2c_traceevent(struct gd32_i2c_priv_s *priv,
                                enum gd32_trace_e event, uint32_t parm)
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 9]
```diff
@@ -0,0 +1,2731 @@
...
+/****************************************************************************
+ * Name: gd32_i2c_setclock
+ *
+ * Description:
+ *   Set the I2C clock
+ *
+ ****************************************************************************/
+
+static void gd32_i2c_setclock(struct gd32_i2c_priv_s *priv,
+                               uint32_t frequency)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:1031]:
```suggestion
static void gd32_i2c_setclock(struct gd32_i2c_priv_s *priv,
                              uint32_t frequency)
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 10]
```diff
@@ -0,0 +1,2731 @@
...
+
+          /* Set condition to activate msg handling */
+
+          priv->dcnt = -1;
+
+#ifndef CONFIG_I2C_POLLED
+          /* Restart ISR by setting an interrupt buffer bit */
+
+          gd32_i2c_modifyreg(priv,
+                              GD32_I2C_CTL1_OFFSET, 0, I2C_CTL1_BUFIE);
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:1423]:
```suggestion
          gd32_i2c_modifyreg(priv,
                             GD32_I2C_CTL1_OFFSET, 0, I2C_CTL1_BUFIE);
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 11]
```diff
@@ -0,0 +1,2731 @@
...
+          /* Send address byte with correct 8th bit set
+           * (for writing or reading)
+           * Transmission happens after having written to the data register
+           * GD32_I2C_DATA
+           */
+
+          gd32_i2c_putreg(priv, GD32_I2C_DATA_OFFSET,
+                           (priv->flags & I2C_M_TEN) ?
+                           0 : ((priv->msgv->addr << 1) |
+                           (priv->flags & I2C_M_READ)));
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:1393]:
```suggestion
          gd32_i2c_putreg(priv, GD32_I2C_DATA_OFFSET,
                          (priv->flags & I2C_M_TEN) ?
                          0 : ((priv->msgv->addr << 1) |
                          (priv->flags & I2C_M_READ)));
```

File: arch/arm/src/gd32f4/gd32f4xx_i2c.c [hunk 12]
```diff
@@ -0,0 +1,2731 @@
...
+#endif
+        }
+
+      i2cinfo("Switch to new message\n");
+
+      /* Get current message to process data and copy to private structure */
+
+      priv->ptr           = priv->msgv->buffer;   /* Copy buffer to private struct */
+      priv->dcnt          = priv->msgv->length;   /* Set counter of current msg length */
+      priv->flags         = priv->msgv->flags;    /* Copy flags to private struct */
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_i2c.c:1319]: maybe some extra spaces before `=` can be removed here?

Comment (GD32-MCU) [reply arch/arm/src/gd32f4/gd32f4xx_i2c.c]: Thanks，I will modify according to comments.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**